### PR TITLE
Ports - Amy W and Mina

### PR DESCRIPTION
**Pull request by minams**
# Ride Share
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
How did getting up to speed on the existing codebase go? If it went well, what worked? If it didn't go well, what will you do differently next time? | It was a little confusing at first to organize all the information since everything was referencing everything else.
What _inheritance_ relations exist between classes? | CSVRecords is the super class that feeds into all the classes except for TripDispatcher.
What _composition_ relations exist between classes? | Trip contains Driver and Passenger.  TripDispatcher will contain information about Trip, Driver, and Passenger.
Describe a decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | We decided for trips that are less than 1.65, the driver will not  owe money for that trip.  The company will not take the fee.  We considered taking out the whole trip but decided to keep a record of that trip.
Give an example of a _template method_ that you implemented for this assignment | We implemented the pathway inherited from the CSVRecords file to be able to access the directory.
Give an example of a _nominal test_ that you wrote for this assignment | That the driver's average rating computed correctly.
Give an example of an _edge case test_ that you wrote for this assignment | That the rating, cost, and end_time were nil for a requested trip.
What is a concept that you gained more clarity on as you worked on this assignment | Importance of keeping data types consistent and being able to recognize when we changed these data types.

### CODE REVIEW

File: specs/trip_dispatch_spec.rb [hunk 1]
```diff
@@ -122,4 +121,61 @@ def build_test_dispatcher
...
+
+    #   expect(@dispatcher.request_trip(passenger_id).start_time).must_equal Time.now
+    # end
+
+    it "will look for nil values for end_time, cost, and rating" do
+      passenger_id = 8
+
+      expect(@dispatcher.request_trip(passenger_id).end_time).must_be_nil
+      expect(@dispatcher.request_trip(passenger_id).cost).must_be_nil
+      expect(@dispatcher.request_trip(passenger_id).rating).must_be_nil
```
Comment (CheezItMan) [inline specs/trip_dispatch_spec.rb:178]: This is crashing because your method is returning an array and array has no `rating` method.

File: lib/trip_dispatcher.rb [hunk 2]
```diff
@@ -27,14 +34,37 @@ def inspect
               #{passengers.count} passengers>"
     end
 
+    def request_trip(passenger_id)
+      passenger = find_passenger(passenger_id)
+
+      @drivers.each do |driver|
```
Comment (CheezItMan) [inline lib/trip_dispatcher.rb:40]: If you don't find a driver here, you're ending up implicitly returning the @drivers array, which is **not** what you want.  Instead you should have a `raise` after the loop, or return something sensible.

File: specs/trip_dispatch_spec.rb [hunk 3]
```diff
@@ -122,4 +121,61 @@ def build_test_dispatcher
...
+    it "assign driver to requested trip" do
+      passenger_id = 8
+
+      expect(@dispatcher.request_trip(passenger_id).driver_id).must_equal 2
+    end
+
+    # it "will return current time for start time for new requested trip" do
+    #   passenger_id = 8
+
+    #   expect(@dispatcher.request_trip(passenger_id).start_time).must_equal Time.now
```
Comment (CheezItMan) [inline specs/trip_dispatch_spec.rb:170]: You commented this out, and I would suggest that you instead of testing for equals test to see that the times are close, because some time will have past between requesting a trip and getting to the test.

File: lib/trip.rb [hunk 4]
```diff
@@ -1,33 +1,55 @@
...
         @passenger = passenger
         @passenger_id = passenger.id
-
       elsif passenger_id
         @passenger_id = passenger_id
+      else
+        raise ArgumentError, "Passenger or passenger_id is required"
+      end
 
+      if start_time == nil
```
Comment (CheezItMan) [inline lib/trip.rb:33]: This if statement doesn't make sense.  You're setting start_time to nil if start_time is nil

This would be better as:


```ruby
@start_time = Time.parse(start_time) unless start_time.nil?
```

File: lib/trip.rb [hunk 5]
```diff
@@ -40,22 +62,32 @@ def inspect
...
       @passenger = passenger
       passenger.add_trip(self)
     end
 
+    def connect_driver(driver)
+      @driver = driver
+      driver.add_trip(self)
+    end
+
+    def duration
```
Comment (CheezItMan) [inline lib/trip.rb:75]: What about for incomplete trips?

File: lib/driver.rb [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    def add_trip(trip)
+      @trips << trip
+    end
+
+    # what is this driver's average rating?
+    def average_rating
+      return 0 if @trips.length == 0
+
+      ratings = []
+      @trips.each do |trip|
```
Comment (CheezItMan) [inline lib/driver.rb:32]: What about incomplete trips?

File: lib/driver.rb [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+        ratings << trip.rating
+      end
+      average_rating = ratings.sum.to_f / ratings.length.to_f
+      return average_rating
+    end
+
+    # calculates driver's total revenue across all trips. each driver gets 80% of trip cost after a fee of 1.65 is deducted
+    def total_revenue # need to somehow figure out how to indicate length of array
+      cost_per_trip = []
```
Comment (CheezItMan) [inline lib/driver.rb:41]: Good work handling the 1.65 fee, but what about incomplete trips where cost is still `nil`.

**Comment by CheezItMan** [comment]
# Ride Share
## What We're Looking For

Feature|Feedback
:------------- | :-------------
**Baseline** |
Used Git Regularly | Reasonable number of commits and mostly good commit messages, but don't use wave numbers in your commit messages.  Instead describe the functionality you've added.
Answer comprehension questions | Check, well done
**Wave 1** |
Appropriate use of Ruby's `Time` | Check
`Trip` has a helper method to calculate duration | Check
`Passenger` methods to calculate total cost and total time | Check
Complete tests for new behavior | Check
**Wave 2** |
`Driver` inherits from `CsvRecord`, and implements `from_csv` | Check
`Driver`s and `Trip`s are connected by `TripDispatcher` | Check
`Driver` has method to calculate average rating | Check
`Driver` has a method to calculate total revenue | Check, good work handling the 1.65 fee
**Wave 3** |
`TripDispatcher` has a new method to create trips | Check, but see my notes for an error case.
Trip creation logic is correctly implemented | See my inline notes
New `Trip` is connected to the appropriate `Driver` and `Passenger` | Check
Tests for `request_trip` | You're missing tests for when all drivers are busy
Wave 1 and 2 methods ignore in-progress trips | **NOPE**
Tests for Wave 1 and 2 methods w/ in-progress trips | **NOPE**
**Overall** | Not bad, but you missed some edge cases both in your testing and application code.  See my inline notes.  I highly suggest in future projects you put an emphasis on doing through testing.  That is a growth areas I'd like you both to focus on.
